src/main/utility/childpid_watcher.rs

```diff
@@ -0,0 +1,465 @@
+};
+use nix::unistd::close;
+use nix::unistd::Pid;
+use std::collections::HashMap;
+use std::convert::{TryFrom, TryInto};
+use std::os::unix::io::RawFd;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::thread;
+
+pub struct ChildPidWatcher {
+    inner: Arc<Mutex<Inner>>,
+    epoll: std::os::unix::io::RawFd,
+}
+
+pub type WatchHandle = u64;
+
+enum Command {
+    RunCallbacks(Pid),
+    Finish,
+}
+
+struct Inner {
+    // Next unique handle ID.
+    next_handle: WatchHandle,
+    // Pending commands for watcher thread.
+    commands: Vec<Command>,
+    // For each monitored pid, the fd obtained via `pidfd_open`.
+    pidfds: HashMap<Pid, RawFd>,
+    // Registered callbacks.
+    callbacks: HashMap<Pid, HashMap<WatchHandle, Box<dyn Send + FnOnce(Pid)>>>,
+    // event_fd used to notify watcher thread via epoll. Calling thread writes a
+    // single byte, which the watcher thread reads to reset.
+    command_notifier: RawFd,
+    thread_handle: Option<thread::JoinHandle<()>>,
+}
+
+impl Inner {
+    fn send_command(&mut self, cmd: Command) {
+        self.commands.push(cmd);
+        nix::unistd::write(self.command_notifier, &1u64.to_ne_bytes()).unwrap();
+    }
+
+    fn unwatch_pid(&mut self, epoll: RawFd, pid: Pid) {
+        self.callbacks.remove(&pid);
+        if let Some(pidfd) = self.pidfds.remove(&pid) {
+            epoll_ctl(epoll, EpollOp::EpollCtlDel, pidfd, None).unwrap();
+            close(pidfd).unwrap();
+        }
+    }
+
+    fn run_callbacks_for_fd(&mut self, epoll: RawFd, pid: Pid) {
+        if let Some(callback_map) = self.callbacks.remove(&pid) {
+            for (_handle, cb) in callback_map {
+                cb(pid)
+            }
+        }
+        self.unwatch_pid(epoll, pid);
+    }
+}
+
+// Panics if pid doesn't exist at all.
+fn is_zombie(pid: Pid) -> bool {
+    let stat_name = format!("/proc/{}/stat", pid.as_raw());
+    let contents = std::fs::read_to_string(stat_name).unwrap();
+    contents.contains(") Z")
+}
+
+impl ChildPidWatcher {
+    /// Get a reference to the global singleton watcher.
+    pub fn new() -> Self {
+        let epoll = epoll_create1(EpollCreateFlags::empty()).unwrap();
+        let command_notifier =
+            nix::sys::eventfd::eventfd(0, nix::sys::eventfd::EfdFlags::EFD_NONBLOCK).unwrap();
+        let mut event = EpollEvent::new(EpollFlags::EPOLLIN, 0);
+        epoll_ctl(
+            epoll,
+            EpollOp::EpollCtlAdd,
+            command_notifier,
+            Some(&mut event),
+        )
+        .unwrap();
+        let watcher = ChildPidWatcher {
+            inner: Arc::new(Mutex::new(Inner {
+                next_handle: 1,
+                pidfds: HashMap::new(),
+                callbacks: HashMap::new(),
+                commands: Vec::new(),
+                command_notifier,
+                thread_handle: None,
+            })),
+            epoll,
+        };
+        let thread_handle = {
+            let inner = watcher.inner.clone();
```

Comment:
Done.

That's one danger of the otherwise very nice vscode+rls type annotations - in my IDE I see `let inner: Arc<Mutex<Inner>> = watcher.inner.clone()`, which makes it clear to *me* what's happening, but not so much for a reader without those annotations.